File: backend/app/schemas/portal.py
"""Pydantic schemas for customer portal API."""

from __future__ import annotations

import uuid
from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel, EmailStr, Field


# --- Auth ---

class PortalRefreshRequest(BaseModel):
    refresh_token: str = Field(min_length=10)


class PortalLoginRequest(BaseModel):
    tenant_slug: str = Field(min_length=1, max_length=100)
    email: EmailStr
    password: str = Field(min_length=8, max_length=128)


class PortalTokenResponse(BaseModel):
    access_token: str
    token_type: str = "bearer"
    expires_in: int
    refresh_token: str | None = None
    tenant_slug: str
    tenant_name: str


class PortalUserResponse(BaseModel):
    id: uuid.UUID
    email: str
    full_name: str
    job_title: str | None
    phone: str | None
    contact_id: uuid.UUID
    company_id: uuid.UUID | None
    company_name: str | None
    tenant_slug: str
    tenant_name: str

    model_config = {"from_attributes": True}


class PortalProfileUpdate(BaseModel):
    full_name: str | None = Field(default=None, min_length=1, max_length=255)
    job_title: str | None = Field(default=None, max_length=150)
    phone: str | None = Field(default=None, max_length=50)


class PortalCreateUserRequest(BaseModel):
    contact_id: uuid.UUID
    email: EmailStr
    password: str = Field(min_length=8, max_length=128)
    full_name: str | None = Field(default=None, max_length=255)


# --- Dashboard ---

class PortalKpiWidget(BaseModel):
    key: str
    label: str
    value: str | int
    hint: str | None = None


class PortalDashboardResponse(BaseModel):
    kpis: list[PortalKpiWidget]
    open_deals: list[PortalDealSummary]
    upcoming_meetings: list[PortalMeetingSummary]
    recent_documents: list[PortalDocumentSummary]
    recent_activities: list[PortalTimelineItem]
    announcements: list[PortalAnnouncementSummary]
    unread_notifications: int
    pending_signatures: int
    open_tickets: int
    outstanding_payments: int


# --- Deals ---

class PortalDealSummary(BaseModel):
    id: uuid.UUID
    title: str
    stage: str
    stage_label: str
    value: Decimal | None
    currency: str
    probability: int
    expected_close_date: date | None
    updated_at: datetime


class PortalDealDetail(PortalDealSummary):
    description: str | None
    timeline: list[PortalTimelineItem]


# --- Documents ---

class PortalDocumentSummary(BaseModel):
    id: uuid.UUID
    name: str
    status: str
    mime_type: str
    size_bytes: int
    current_version: int
    updated_at: datetime
    deal_id: uuid.UUID | None = None


class PortalDocumentDetail(PortalDocumentSummary):
    description: str | None
    versions: list[PortalDocumentVersion]


class PortalDocumentVersion(BaseModel):
    version_number: int
    filename: str
    size_bytes: int
    created_at: datetime


class PortalDocumentUploadMeta(BaseModel):
    name: str = Field(min_length=1, max_length=500)
    description: str | None = None
    deal_id: uuid.UUID | None = None


# --- Meetings ---

class PortalMeetingSummary(BaseModel):
    id: uuid.UUID
    title: str
    status: str
    meeting_type: str
    start_datetime: datetime
    end_datetime: datetime
    location: str | None
    meeting_url: str | None


class PortalMeetingRequest(BaseModel):
    title: str = Field(min_length=1, max_length=255)
    description: str | None = None
    preferred_start: datetime
    preferred_end: datetime
    meeting_type: str = "client_meeting"


# --- Support ---

class PortalTicketCreate(BaseModel):
    subject: str = Field(min_length=1, max_length=255)
    description: str = Field(min_length=1, max_length=8000)
    priority: str = "medium"
    category: str = "general"


class PortalTicketReplyCreate(BaseModel):
    body: str = Field(min_length=1, max_length=8000)


class PortalTicketReplyResponse(BaseModel):
    id: uuid.UUID
    author_type: str
    author_name: str
    body: str
    created_at: datetime


class PortalTicketSummary(BaseModel):
    id: uuid.UUID
    subject: str
    status: str
    priority: str
    category: str
    created_at: datetime
    updated_at: datetime
    reply_count: int


class PortalTicketDetail(PortalTicketSummary):
    description: str
    replies: list[PortalTicketReplyResponse]


# --- Knowledge & announcements ---

class PortalAnnouncementSummary(BaseModel):
    id: uuid.UUID
    title: str
    body: str
    published_at: datetime | None
    created_at: datetime


class PortalKnowledgeSummary(BaseModel):
    id: uuid.UUID
    title: str
    slug: str
    summary: str | None
    category: str
    view_count: int


class PortalKnowledgeDetail(PortalKnowledgeSummary):
    body: str


# --- Notifications & timeline ---

class PortalNotificationItem(BaseModel):
    id: uuid.UUID
    notification_type: str
    title: str
    body: str | None
    link: str | None
    is_read: bool
    created_at: datetime


class PortalTimelineItem(BaseModel):
    id: str
    event_type: str
    title: str
    detail: str | None
    occurred_at: datetime
    entity_type: str | None = None
    entity_id: uuid.UUID | None = None


# --- Invoices ---

class PortalInvoiceSummary(BaseModel):
    id: uuid.UUID
    invoice_number: str
    amount: Decimal
    currency: str
    status: str
    due_date: date | None
    deal_id: uuid.UUID | None
    document_id: uuid.UUID | None


# --- AI ---

class PortalAiChatMessage(BaseModel):
    role: str = Field(pattern="^(user|assistant)$")
    content: str = Field(min_length=1, max_length=32000)


class PortalAiChatRequest(BaseModel):
    messages: list[PortalAiChatMessage] = Field(min_length=1, max_length=30)

File: frontend/types/portal.ts
export interface PortalUser {
  id: string;
  email: string;
  full_name: string;
  job_title: string | null;
  phone: string | null;
  contact_id: string;
  company_id: string | null;
  company_name: string | null;
  tenant_slug: string;
  tenant_name: string;
}

export interface PortalKpi {
  key: string;
  label: string;
  value: string | number;
  hint?: string | null;
}

export interface PortalDeal {
  id: string;
  title: string;
  stage: string;
  stage_label: string;
  value: string | null;
  currency: string;
  probability: number;
  expected_close_date: string | null;
  updated_at: string;
}

export interface PortalDealDetail extends PortalDeal {
  description: string | null;
  timeline: PortalTimelineItem[];
}

export interface PortalDocument {
  id: string;
  name: string;
  status: string;
  mime_type: string;
  size_bytes: number;
  current_version: number;
  updated_at: string;
  deal_id?: string | null;
}

export interface PortalDocumentDetail extends PortalDocument {
  description: string | null;
  versions: { version_number: number; filename: string; size_bytes: number; created_at: string }[];
}

export interface PortalMeeting {
  id: string;
  title: string;
  status: string;
  meeting_type: string;
  start_datetime: string;
  end_datetime: string;
  location: string | null;
  meeting_url: string | null;
}

export interface PortalTicket {
  id: string;
  subject: string;
  status: string;
  priority: string;
  category: string;
  created_at: string;
  updated_at: string;
  reply_count: number;
}

export interface PortalTicketReply {
  id: string;
  author_type: string;
  author_name: string;
  body: string;
  created_at: string;
}

export interface PortalTicketDetail extends PortalTicket {
  description: string;
  replies: PortalTicketReply[];
}

export interface PortalAnnouncement {
  id: string;
  title: string;
  body: string;
  published_at: string | null;
  created_at: string;
}

export interface PortalKnowledgeSummary {
  id: string;
  title: string;
  slug: string;
  summary: string | null;
  category: string;
  view_count: number;
}

export interface PortalKnowledgeArticle extends PortalKnowledgeSummary {
  body: string;
}

export interface PortalNotification {
  id: string;
  notification_type: string;
  title: string;
  body: string | null;
  link: string | null;
  is_read: boolean;
  created_at: string;
}

export interface PortalTimelineItem {
  id: string;
  event_type: string;
  title: string;
  detail: string | null;
  occurred_at: string;
  entity_type?: string | null;
  entity_id?: string | null;
}

export interface PortalInvoice {
  id: string;
  invoice_number: string;
  amount: string;
  currency: string;
  status: string;
  due_date: string | null;
  deal_id: string | null;
  document_id: string | null;
}

export interface PortalOrganization {
  tenant_name: string;
  company_id: string | null;
  company_name: string | null;
  contact_name: string;
  contact_email: string;
  contact_phone: string | null;
}

export interface PortalProfileUpdate {
  full_name?: string;
  job_title?: string | null;
  phone?: string | null;
}

export interface PortalMeetingRequest {
  title: string;
  description?: string;
  preferred_start: string;
  preferred_end: string;
  meeting_type?: string;
}

export interface PortalDashboard {
  kpis: PortalKpi[];
  open_deals: PortalDeal[];
  upcoming_meetings: PortalMeeting[];
  recent_documents: PortalDocument[];
  recent_activities: PortalTimelineItem[];
  announcements: PortalAnnouncement[];
  unread_notifications: number;
  pending_signatures: number;
  open_tickets: number;
  outstanding_payments: number;
}
